File: kdtree/kdtree.cpp
#include "kdtree.h"

double Util::distance(vector<double> x, vector<double> y) {
    double distance = 0;
    int dim = x.size();
    for (int i = 0; i < dim; i ++) {
        distance += pow(x[i] - y[i], 2);
    }
    return sqrt(distance);
}


KdNode::KdNode(vector<vector<double> > coordinateVec, int target, int splitDim) {
    this->splitDim = coordinateVec.size() == 1? -1: splitDim;
    this->coordinate = coordinateVec[target];
    double compare = 0;
    int nodeCount = coordinateVec.size();
    for (int i = 0; i < nodeCount; i ++) {
        compare = coordinateVec[i][splitDim] - this->coordinate[splitDim];
        if (compare > 0) {
            this->rightVec.push_back(coordinateVec[i]);
        } else if (i != target) {
            this->leftVec.push_back(coordinateVec[i]);
        }
    }
}

KdNode::~KdNode(void) {
    this->coordinate.clear();
    if (this->left != NULL) {
        delete this->left;
        this->left = NULL;
    }
    if (this->right != NULL) {
        delete this->right;
        this->right = NULL;
    }
}

int KdNode::getSplitDim(void) {
    return this->splitDim;
}

vector<vector<double> > KdNode::getLeftVec(void) {
    return this->leftVec;
}

vector<vector<double> > KdNode::getRightVec(void) {
    return this->rightVec;
}

void KdNode::done(void) {
    this->leftVec.clear();
    this->rightVec.clear();
}

double KdNode::distance(vector<double> coordinate) {
    return Util::distance(this->coordinate, coordinate);
}

KdNode* KdTree::createNode(vector<vector<double> > coordinateVec) {
    vector<vector<double> > vecs = coordinateVec;
    int nodeCount = vecs.size();
    // if coordinateVec is empty, return NULL
    if (nodeCount == 0) {
        return NULL;
    }
    int dim = vecs[0].size();
    int splitDim = 0;
    vector<double> var(dim, 0);
    vector<double> avg(dim, 0);
    vector<double> order(nodeCount, 0);
    map<double, int> keyMap;
    double maxVar = 0;
    int midPos = nodeCount / 2;
    for (int i = 0; i < nodeCount; i ++) {
        for (int j = 0; j < dim; j ++) {
            avg[j] += vecs[i][j];
        }
    }
    for (int j = 0; j < dim; j ++) {
        avg[j] /= nodeCount;
    }
    for (int i = 0; i < nodeCount; i ++) {
        for (int j = 0; j < dim; j ++) {
            var[j] += pow(vecs[i][j] - avg[j], 2);
        }
    }

    for (int j = 0; j < dim; j ++) {
        if (var[j] > maxVar) {
            maxVar = var[j];
            splitDim = j;
        }
    }

    for (int i = 0; i < nodeCount; i ++) {
        order[i] = vecs[i][splitDim];
        keyMap.insert(pair<double, int>(order[i], i));
    }
    sort(order.begin(), order.end());
    int idx = keyMap.find(order[midPos])->second;
    return new KdNode(vecs, idx, splitDim);
}

KdTree::KdTree(vector<vector<double> > coordinateVec) {
    stack<KdNode*> nodeStack;
    bool leftFlag = true;
    KdNode* root = this->createNode(coordinateVec);
    KdNode* node = this->createNode(root->getLeftVec());
    this->root = root;
    nodeStack.push(root);

    while (true) {
        if (leftFlag) {
            root->left = node;
        } else {
            root->right = node;
            root->done();
        }

        if (node == NULL) {
            if (nodeStack.empty()) {
                break;
            }
            root = nodeStack.top();
            nodeStack.pop();
            leftFlag = false;
            node = this->createNode(root->getRightVec());
        } else {
            root = node;
            nodeStack.push(root);
            leftFlag = true;
            node = this->createNode(root->getLeftVec());
        }
    }
}

KdTree::~KdTree(void) {
    delete this->root;
    this->root = NULL;
}

vector<double> KdTree::search(vector<double> coordinate) {
    stack<KdNode*> nodeStack;
    stack<bool> leftVisitStack;
    KdNode* node = this->root;
    KdNode* nearest = NULL;
    int splitDim = 0;
    bool leftVisit = false;
    bool trace = false;
    bool flag = true;
    double distance = 0.0;
    double tDis = 0.0;

    while (flag) {
        
        if (node == NULL) {
            node = nodeStack.top();
            nodeStack.pop();
            leftVisitStack.pop();
            trace = true;
        } else {
            splitDim = node->getSplitDim();
            if (splitDim != -1) {
                nodeStack.push(node);
                if (coordinate[splitDim] <= node->coordinate[splitDim]) {
                    node = node->left;
                    leftVisitStack.push(true);
                } else {
                    node = node->right;
                    leftVisitStack.push(false);
                }
            } else {
                trace = true;
            }
        }

        if (trace) {
            tDis = node->distance(coordinate);
            if (nearest == NULL || tDis < distance) {
                distance = tDis;
                nearest = node;
            }
        }

        while (trace) {
            if (nodeStack.empty()) {
                flag = false;
                break;
            }
            node = nodeStack.top();
            nodeStack.pop();
            leftVisit = leftVisitStack.top();
            leftVisitStack.pop();
            tDis = node->distance(coordinate);
            if (distance > tDis) {
                distance = tDis;
                nearest = node;
            }
            splitDim = node->getSplitDim();
            if (pow(distance, 2) > pow(node->coordinate[splitDim] - coordinate[splitDim], 2)) {
                node = leftVisit? node->right: node->left;
                trace = false;
            }
        }
    }
    // node = NULL;
    return nearest->coordinate;
}

int main() {
    vector<vector<double> > coordinateVec;
    string data;
    ifstream reader("points.txt");
    char* p;
    const char* splitChars = ",";
    while (getline(reader, data)) {
        vector<double> coordinate;
        p = strtok(const_cast<char*>(data.c_str()), splitChars);
        while (p != NULL) {
            coordinate.push_back(atof(p));
            p = strtok(NULL, splitChars);
        }
        coordinateVec.push_back(coordinate);
    }
    reader.close();

    delete[] p;

    vector<double> point;
    point.push_back(2);
    point.push_back(4.5);

    KdTree* kdT = new KdTree(coordinateVec);
    vector<double> nearest = kdT->search(point);
    vector<double>::iterator it = nearest.begin();
    cout << "neareast: ";
    while (it != nearest.end()) {
        printf("%.2lf ", *it);
        it ++;
    }
    cout << "\ndistance: " << Util::distance(nearest, point) << endl;

    delete kdT;
    kdT = NULL;
    return 0;
}

File: kdtree/kdtree.py
#!/usr/bin/python
#coding: UTF-8
import os
import math

class Util:

    @staticmethod
    def distance(x, y):
        return math.sqrt(sum([(x[i] - y[i]) ** 2 for i in range(len(x))]))


class KdNode:

    def __init__(self, coordinateVec, target, splitDim):
        self.left = None
        self.right = None
        self.coordinate = coordinateVec[target]
        self.__splitDim = splitDim
        self.__leftVec = []
        self.__rightVec = []

        if len(coordinateVec) == 1:
            self.__splitDim = -1        

        for i, v in enumerate(coordinateVec):
            compare = v[splitDim] - self.coordinate[splitDim]
            if compare > 0:
                self.__rightVec.append(v)
            elif i != target:
                self.__leftVec.append(v)


    def getSplitDim(self):
        return self.__splitDim


    def getLeftVec(self):
        return self.__leftVec


    def getRightVec(self):
        return self.__rightVec


    def done(self):
        self.__leftVec = []
        self.__rightVec = []


    def distance(self, coordinate):
        return Util.distance(self.coordinate, coordinate)


class KdTree:

    def __init__(self, coordinateVec):
        nodeStack = []
        leftFlag = True
        root = self.__createNode(coordinateVec)
        node = self.__createNode(root.getLeftVec())
        
        self.__root = root

        nodeStack.append(root)
        while True:
            if leftFlag:
                root.left = node
            else:
                root.right = node
                root.done()

            if not node:
                if not nodeStack:
                    break
                root = nodeStack.pop()
                leftFlag = False
                node = self.__createNode(root.getRightVec())
            else:
                root = node
                nodeStack.append(root)
                leftFlag = True
                node = self.__createNode(root.getLeftVec())


    def __createNode(self, coordinateVec):
        size = len(coordinateVec)
        if size == 0:
            return None

        dim = len(coordinateVec[0])
        splitDim = 0
        maxVar = 0
        midPos = size // 2

        trans = [[r[i] for r in coordinateVec] for i in range(dim)]
        avg = [sum(r) / size for r in trans]
        var = [sum([(k - avg[i]) ** 2 for k in r]) for r in trans] 

        for i, v in enumerate(var):
            if v > maxVar:
                maxVar = v
                splitDim = i

        order = sorted(trans[splitDim])
        keyMap = {v: i for i, v in enumerate(trans[splitDim])}
        return KdNode(coordinateVec, keyMap[order[midPos]], splitDim)
        


    def search(self, coordinate):
        nodeStack = []
        leftVisitStack = []
        node = self.__root
        nearest = None
        trace = False
        flag = True

        while flag:
            if not node:
                node = nodeStack.pop()
                leftVisitStack.pop()
                trace = True
            else:
                splitDim = node.getSplitDim()
                if splitDim != -1:
                    nodeStack.append(node)
                    
                    if coordinate[splitDim] <= node.coordinate[splitDim]:
                        node = node.left
                        leftVisitStack.append(True)
                    else:
                        node = node.right
                        leftVisitStack.append(False)
                else:
                    trace = True

            if trace:
                tDis = node.distance(coordinate)
                if not nearest or tDis < distance:
                    distance = tDis
                    nearest = node

            while trace:
                if not nodeStack:
                    flag = False
                    break

                node = nodeStack.pop()
                leftVisit = leftVisitStack.pop()

                tDis = node.distance(coordinate)
                if distance > tDis:
                    distance = tDis
                    nearest = node

                splitDim = node.getSplitDim()
                if distance ** 2 > (node.coordinate[splitDim] - coordinate[splitDim]) ** 2:
                    if leftVisit:
                        node = node.right
                    else:
                        node = node.left
                    trace = False

        return nearest.coordinate


if __name__ == '__main__':
    coordinateVec = []
    with open('points.txt') as f:
        coordinateVec = [[float(k) for k in l.strip().split(',')] for l in f if l]

    point = [2, 4.5]
    
    kdT = KdTree(coordinateVec)
    nearest = kdT.search(point)
    
    print(','.join([str(k) for k in nearest]))
    print('distance: {0}'.format(Util.distance(nearest, point)))
